keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1392 @@
+crashes or restarts (shortly after API server becoming and staying unavailable),
+the `lastTransitionTime` fields may be inaccurate. This is described in the
+section [above](#unavailability-of-api-server-or-etcd-along-with-kubelet-restart).
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+None so far
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+SLOs are not applicable to pod status fields. Overall Kubernetes node level SLOs
+may leverage this feature.
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+The main drawback associated with the new pod sandbox conditions involves a
+slight potential increase in calls to the API Server from Kubelet to patch
+`status` = `true` for the new conditions in a pod's status. Typically, this
+would involve two extra patch calls for pod status in the lifetime of most pods
+(if the status manager does not batch them with other pod status updates): one
+when pod sandbox creation actually starts and the other when pod sandbox
+creation completes successfully. However, there could be a higher number of
+patch calls to API Server if the pod sandbox environment (like a microvm) starts
+successfully and then crashes in a re-start loop.
+
+Caching of updates to pod status by the pod status manager and batching pod
+status updates (which is already in place) can help mitigate frequent patch
+calls to API server.
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Dedicated fields or annotations for the pod sandbox creation timestamps
```

Comment:
Would the traces enabled by https://github.com/kubernetes/enhancements/issues/2831 be another alternative that could address this? Traces will give us even more info during the pod lifecycle beyond the sandbox creation data.

Reply:
Data in the pod sandbox conditions within each pod enables:
[1] A metric collection service to analyze, slice and aggregate the data across (a) arbitrary labels and annotations in a pod denoting classes of workloads (b) fields like storage classes of PVCs mounted by the pod, runtimeclass, etc.
[2] A controller to act on the data (as described [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R228))

From a pure aggregated metric collection perspective, tracing could be an alternative. However more specialized classification/consumption of the data requires Kubelet to surface the raw data on a per-pod basis (as the existing pod status conditions already do). I will add this in the Alternatives section.

Reply:
Clarified [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R1518)